chromium/allocator.go

```diff
@@ -1,69 +1,140 @@
 	"github.com/grafana/xk6-browser/common"
 )
 
-// Ensure Allocator implements the browser.Allocator interface.
-var _ allocation.Allocator = &Allocator{}
-
+// Allocator provides facilities for finding, running, and interacting with a Chromium browser.
 type Allocator struct {
-	execPath  string
-	initFlags map[string]interface{}
-	initEnv   []string
-	tempDir   string
+	execPath      string                 // path to the Chromium executable
+	initFlags     map[string]interface{} // CLI flags to pass to the Chromium executable
+	initEnv       []string               // environment variables to pass to the Chromium executable
+	tempDir       string                 // path for storing the extension and user specific data
+	removeTempDir func()                 // for removing the tempDir
+}
+
+// NewAllocator returns a new Allocator with a path to a Chromium executable.
+func NewAllocator(flags map[string]interface{}, env []string) *Allocator {
+	return &Allocator{
+		execPath:      findExecPath(),
+		initFlags:     flags,
+		initEnv:       env,
+		tempDir:       "",
+		removeTempDir: func() {},
+	}
+}
+
+// Allocate starts a new Chromium browser process and returns it.
+func (a *Allocator) Allocate(
+	ctx context.Context, launchOpts *common.LaunchOptions,
+) (_ *common.BrowserProcess, rerr error) {
+	args, err := a.prepareArgs()
+	if err != nil {
+		return nil, fmt.Errorf("cannot prepare args: %w", err)
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	cmd := exec.CommandContext(ctx, a.execPath, args...) //nolint:gosec
+	killAfterParent(cmd)
 
-	wg sync.WaitGroup
+	defer func() {
+		if rerr != nil {
+			cancel()
+			a.removeTempDir()
+		}
+	}()
 
-	combinedOutputWriter io.Writer
+	// Pipe stderr to stdout
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, fmt.Errorf("cannot pipe stdout: %w", err)
+	}
+	cmd.Stderr = cmd.Stdout
+
+	// Set up environment variable for process
+	if len(a.initEnv) > 0 {
+		cmd.Env = append(os.Environ(), a.initEnv...)
+	}
+
+	// We must start the cmd before calling cmd.Wait, as otherwise the two
+	// can run into a data race.
+	if err := cmd.Start(); err != nil {
+		return nil, fmt.Errorf("cannot start browser executable: %w", err)
+	}
+	if ctx.Err() != nil {
+		return nil, fmt.Errorf("context err after command start: %w", ctx.Err())
+	}
+
+	go func() {
+		_ = cmd.Wait()
+		a.removeTempDir()
+	}()
+
+	ctxTimeout, cancel := context.WithTimeout(ctx, launchOpts.Timeout)
+	defer cancel()
+
+	wsURL, err := a.parseWebsocketURL(ctxTimeout, stdout)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse websocket url: %w", err)
+	}
+	_ = stdout.Close()
+
+	return common.NewBrowserProcess(ctx, cancel, cmd.Process, wsURL, a.tempDir), nil
+}
+
+// prepareArgs for launching a Chrome browser with the args.
+func (a *Allocator) prepareArgs() (args []string, err error) {
+	// use the provided directory or create a temporary one.
+	a.tempDir, a.removeTempDir, err = makeUserDataDir(a.tempDir, a.initFlags["user-data-dir"])
```

Comment:
Makes sense 👍 

I didn't want to complicate the `Allocate` and made another type called `DataStore` in the chromium pkg. We can use it for other upcoming features instead of complicating the allocator. By the way, I called it `DataStore` instead of `UserDataStore` because this directory can store both the extension and user-specific data. See: ae6adad. WDYT?

If this is an OK change, I'll create tests for `parseArgs` and `parseWebsocketURL` as well.

Reply:
I haven't tested it, but it looks like an improvement :thumbsup: As long as it's created outside of `prepareArgs` :wink: `Allocate()` seems like the best place for that.

Reply:
What about 76ff7b5?

Reply:
Sure :+1: Though I'm not sure what we need the first argument to `Make()` for if it's always empty?

Anyway, please re-request a review when you're done and I'll take another look.

Reply:
Yeah, it's a design flaw that allows the test for errors from `Mkdir` in `Make`:

https://github.com/grafana/xk6-browser/blob/0ee9c74fabb8b7d573017f33ef73eff72c445778/chromium/data_store_test.go#L38-L44